File: seminar-05/stack/cpp/stack.cpp
#include <stdexcept>
#include <iostream>

template<typename T>
class Stack {
private:
    struct Node {
        T data;
        Node* next;

        Node(const T& data, Node* next)
            : data(data), next(next) {}
    };

    Node* top;
    int size_;

public:
    /**
     * Создаёт пустой стек.
     */
    Stack() : top(nullptr), size_(0) {}

    /**
     * Деструктор — освобождает память.
     */
    ~Stack() {
        while (!isEmpty()) {
            pop();
        }
    }

    /**
     * Добавляет элемент на вершину стека.
     */
    void push(const T& data) {
        top = new Node(data, top);
        size_++;
    }

    /**
     * Удаляет и возвращает верхний элемент.
     */
    T pop() {
        if (isEmpty()) {
            throw std::runtime_error("pop from empty stack");
        }
        Node* node = top;
        T data = node->data;
        top = top->next;
        delete node;
        size_--;
        return data;
    }

    /**
     * Возвращает верхний элемент без удаления.
     */
    T peek() const {
        if (isEmpty()) {
            throw std::runtime_error("peek from empty stack");
        }
        return top->data;
    }

    /**
     * Проверяет, пуст ли стек.
     */
    bool isEmpty() const {
        return top == nullptr;
    }

    /**
     * Возвращает размер стека.
     */
    int size() const {
        return size_;
    }
};

File: seminar-05/stack/java/Stack.java
public class Stack<T> {

    /**
     * Узел односвязного списка.
     */
    private static class Node<T> {
        T data;
        Node<T> next;

        Node(T data, Node<T> next) {
            this.data = data;
            this.next = next;
        }
    }

    private Node<T> top; // вершина стека
    private int size;    // количество элементов

    /**
     * Создаёт пустой стек.
     */
    public Stack() {
        top = null;
        size = 0;
    }

    /**
     * Добавляет элемент на вершину стека.
     */
    public void push(T data) {
        top = new Node<>(data, top);
        size++;
    }

    /**
     * Удаляет и возвращает верхний элемент.
     */
    public T pop() {
        if (isEmpty()) {
            throw new IllegalStateException("pop from empty stack");
        }
        T data = top.data;
        top = top.next;
        size--;
        return data;
    }

    /**
     * Возвращает верхний элемент без удаления.
     */
    public T peek() {
        if (isEmpty()) {
            throw new IllegalStateException("peek from empty stack");
        }
        return top.data;
    }

    /**
     * Проверяет, пуст ли стек.
     */
    public boolean isEmpty() {
        return top == null;
    }

    /**
     * Возвращает размер стека.
     */
    public int size() {
        return size;
    }

    @Override
    public String toString() {
        return "Stack(size=" + size +
                ", top=" + (top == null ? "null" : top.data) + ")";
    }
}
